IPython/core/guarded_eval.py

```diff
@@ -573,21 +577,55 @@ def _handle_assign(node: ast.Assign, context: EvaluationContext):
             # Before starred
             for i in range(star_or_last_idx):
                 transient_locals[targets[i].id] = values[i]
+                # Check for self.x assignment
+                if class_transients is not None and hasattr(targets[i], "ctx"):
+                    if (
+                        isinstance(targets[i], ast.Attribute)
+                        and isinstance(targets[i].value, ast.Name)
+                        and targets[i].value.id == "self"
+                    ):
+                        class_transients[targets[i].attr] = values[i]
 
             # Starred if exists
             if starred:
                 end = len(values) - (len(targets) - star_or_last_idx - 1)
                 transient_locals[targets[star_or_last_idx].value.id] = values[
                     star_or_last_idx:end
                 ]
+                if (
+                    class_transients is not None
+                    and isinstance(targets[star_or_last_idx], ast.Attribute)
+                    and isinstance(targets[star_or_last_idx].value, ast.Name)
+                    and targets[star_or_last_idx].value.id == "self"
+                ):
```

Comment:
- This looks quite repetitive, I would create a function to check this condition
- The class instance argument can have a name other than "self". I think we could extract it from the function signature and store the name on the context when entering function definition. Maybe this should be addressed in a separate PR, but in any case this shows the case for extracting this check logic into a single place.